core/src/main/java/org/mule/runtime/core/internal/event/AbstractEventContext.java

```diff
@@ -37,130 +37,147 @@
-        LOGGER.debug(this + " empty response was already completed, ignoring.");
-        return;
-      }
+    if (isResponseDone()) {
+      LOGGER.debug(this + " empty response was already completed, ignoring.");
+      return;
+    }
 
-      if (LOGGER.isDebugEnabled()) {
-        LOGGER.debug(this + " response completed with no result.");
-      }
-      responseProcessor.onComplete();
+    if (LOGGER.isDebugEnabled()) {
+      LOGGER.debug(this + " response completed with no result.");
     }
+    responseDone(right(null));
   }
 
   /**
    * {@inheritDoc}
    */
   @Override
   public final void success(CoreEvent event) {
-    synchronized (this) {
-      if (responseProcessor.isTerminated()) {
-        LOGGER.debug(this + " response was already completed, ignoring.");
-        return;
-      }
+    if (isResponseDone()) {
+      LOGGER.debug(this + " response was already completed, ignoring.");
+      return;
+    }
 
-      if (LOGGER.isDebugEnabled()) {
-        LOGGER.debug(this + " response completed with result.");
-      }
-      responseProcessor.onNext(event);
+    if (LOGGER.isDebugEnabled()) {
+      LOGGER.debug(this + " response completed with result.");
     }
+    responseDone(right(event));
   }
 
   /**
    * {@inheritDoc}
    */
   @Override
   public final Publisher<Void> error(Throwable throwable) {
-    synchronized (this) {
-      if (responseProcessor.isTerminated()) {
-        LOGGER.debug(this + " error response was already completed, ignoring.");
-        return empty();
-      }
+    if (isResponseDone()) {
+      LOGGER.debug(this + " error response was already completed, ignoring.");
+      return empty();
+    }
 
+    if (LOGGER.isDebugEnabled()) {
+      LOGGER.debug(this + " responseDone completed with error.");
+    }
+
+    if (throwable instanceof MessagingException) {
       if (LOGGER.isDebugEnabled()) {
-        LOGGER.debug(this + " response completed with error.");
+        LOGGER.debug(this + " handling messaging exception.");
       }
+      return just((MessagingException) throwable)
+          .flatMapMany(exceptionHandler)
+          .doOnNext(handled -> success(handled))
+          .doOnError(rethrown -> responseDone(left(rethrown)))
+          .materialize()
+          .then()
+          .toProcessor();
+    } else {
+      responseDone(left(throwable));
+      return empty();
+    }
+  }
+
+  private synchronized void responseDone(Either<Throwable, CoreEvent> result) {
+    this.result = result;
+    state = STATE_RESPONSE;
+    tryComplete();
+    onResponseConsumerList.stream().forEach(consumer -> signalConsumerSilently(consumer));
+  }
 
-      if (throwable instanceof MessagingException) {
-        if (LOGGER.isDebugEnabled()) {
-          LOGGER.debug(this + " handling messaging exception.");
+  protected synchronized void tryComplete() {
+    if (state == STATE_RESPONSE && !childContexts.stream().anyMatch(context -> !context.isComplete())) {
```

Comment:
use another operation to avoid the double negation for readability (`noneMatch`?)